File: Algorithms/814_Binary_Tree_Pruning/Solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun pruneTree(root: TreeNode?): TreeNode? = root?.let {
        it.prune().run {
            if (this == 0) null else it
        }
    }

    private fun TreeNode?.prune(): Int = if (this == null) 0 else this.`val` +
            this.left.prune().also { if (it == 0) this.left = null } +
            this.right.prune().also { if (it == 0) this.right = null }
}

File: Algorithms/814_Binary_Tree_Pruning/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pruneTree(self, root: TreeNode) -> TreeNode:
        return root if self.prune(root) != 0 else None

    def prune(self, root: TreeNode) -> int:
        if not root:
            return 0
        else:
            left_sum = self.prune(root.left)
            right_sum = self.prune(root.right)

            if left_sum == 0:
                root.left = None
            if right_sum == 0:
                root.right = None

            return root.val + left_sum + right_sum
